src/Microsoft.OData.Core/ODataBatchReaderStream.cs

```diff
@@ -343,308 +80,58 @@ internal int ReadWithDelimiter(byte[] userBuffer, int userBufferOffset, int coun
-                        if (byteCount > 0)
-                        {
-                            ODataBatchUtils.EnsureArraySize(ref bytesForString, lineBufferSize, byteCount);
-                            Buffer.BlockCopy(this.batchBuffer.Bytes, this.batchBuffer.CurrentReadPosition, bytesForString, lineBufferSize, byteCount);
-                            lineBufferSize += byteCount;
-                        }
-
-                        if (this.underlyingStreamExhausted)
-                        {
-                            // Nothing more to read; stop looping
-                            scanResult = ODataBatchReaderStreamScanResult.Match;
-                            this.batchBuffer.SkipTo(lineEndStartPosition + 1);
-                        }
-                        else
-                        {
-                            this.underlyingStreamExhausted = this.batchBuffer.RefillFrom(this.inputContext.Stream, /*preserveFrom*/ lineEndStartPosition);
-                        }
-
-                        break;
-                    case ODataBatchReaderStreamScanResult.Match:
-                        // We found a line end in the buffer
-                        Debug.Assert(lineEndStartPosition >= this.batchBuffer.CurrentReadPosition, "Line end must be at or after current position.");
-                        Debug.Assert(lineEndEndPosition < this.batchBuffer.CurrentReadPosition + this.batchBuffer.NumberOfBytesInBuffer, "Line end must finish withing buffer range.");
-
-                        byteCount = lineEndStartPosition - this.batchBuffer.CurrentReadPosition;
-                        if (byteCount > 0)
-                        {
-                            ODataBatchUtils.EnsureArraySize(ref bytesForString, lineBufferSize, byteCount);
-                            Buffer.BlockCopy(this.batchBuffer.Bytes, this.batchBuffer.CurrentReadPosition, bytesForString, lineBufferSize, byteCount);
-                            lineBufferSize += byteCount;
-                        }
-
-                        this.batchBuffer.SkipTo(lineEndEndPosition + 1);
-                        break;
-                    default:
-                        throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReaderStream_ReadLine));
-                }
-            }
-
-            Debug.Assert(bytesForString != null, "bytesForString != null");
-
-            return this.CurrentEncoding.GetString(bytesForString, 0, lineBufferSize);
         }
 
         /// <summary>
         /// Ensure that a batch encoding exists; if not, detect it from the first couple of bytes of the stream.
         /// </summary>
-        private void EnsureBatchEncoding()
+        /// <param name="stream">The stream to read.</param>
+        protected void EnsureBatchEncoding(Stream stream)
         {
             // If no batch encoding is specified we detect it from the first bytes in the buffer.
             if (this.batchEncoding == null)

```

Comment:
>if (this.batchEncoding == null) [](start = 12, length = 31)

do we need to be thread safe here?

Reply:
I think we should log an issue for this for further investigation since it is not related to the change here.

Reply:
I think we should log an issue for this for further investigation since it is not related to the change here.
Let me know if you think we need to log an separate issue in github. I would like to focus on porting for this PR.

---
In reply to: [140611087](https://github.com/OData/odata.net/pull/960#discussion_r140611087) [](ancestors = 140611087)